examples/escrow/programs/escrow/src/lib.rs

```diff
@@ -0,0 +1,233 @@
+    pub escrow_account: ProgramAccount<'info, EscrowAccount>,
+    pub token_program: AccountInfo<'info>,
+    pub rent: Sysvar<'info, Rent>,
+}
+
+#[derive(Accounts)]
+pub struct Exchange<'info> {
+    #[account(signer)]
+    pub taker: AccountInfo<'info>,
+    #[account(mut)]
+    pub taker_deposit_token_account: CpiAccount<'info, TokenAccount>,
+    #[account(mut)]
+    pub taker_receive_token_account: CpiAccount<'info, TokenAccount>,
+    #[account(mut)]
+    pub pda_deposit_token_account: CpiAccount<'info, TokenAccount>,
+    #[account(mut)]
+    pub initializer_receive_token_account: CpiAccount<'info, TokenAccount>,
+    #[account(mut)]
+    pub initializer_main_account: AccountInfo<'info>,
+    #[account(mut,
+        constraint = escrow_account.taker_amount <= taker_deposit_token_account.amount,
+        constraint = escrow_account.initializer_deposit_token_account == *pda_deposit_token_account.to_account_info().key,
+        constraint = escrow_account.initializer_receive_token_account == *initializer_receive_token_account.to_account_info().key,
+        constraint = escrow_account.initializer_key == *initializer_main_account.key,
+        close = initializer_main_account
+    )]
+    pub escrow_account: ProgramAccount<'info, EscrowAccount>,
+    pub pda_account: AccountInfo<'info>,
+    pub token_program: AccountInfo<'info>,
+}
+
+#[derive(Accounts)]
+pub struct CancelEscrow<'info> {
+    pub initializer: AccountInfo<'info>,
+    #[account(mut)]
+    pub pda_deposit_token_account: CpiAccount<'info, TokenAccount>,
+    pub pda_account: AccountInfo<'info>,
+    #[account(mut,
+        constraint = escrow_account.initializer_key == *initializer.key,
+        constraint = escrow_account.initializer_deposit_token_account == *pda_deposit_token_account.to_account_info().key,
+        close = initializer
+    )]
+    pub escrow_account: ProgramAccount<'info, EscrowAccount>,
+    pub token_program: AccountInfo<'info>,
+}
+
+#[account]
+pub struct EscrowAccount {
+    pub initializer_key: Pubkey,
+    pub initializer_deposit_token_account: Pubkey,
+    pub initializer_receive_token_account: Pubkey,
+    pub initializer_amount: u64,
+    pub taker_amount: u64,
+}
+
+impl<'a, 'b, 'c, 'info> From<&mut InitializeEscrow<'info>>
+    for CpiContext<'a, 'b, 'c, 'info, SetAuthority<'info>>
+{
+    fn from(
+        accounts: &mut InitializeEscrow<'info>,
+    ) -> CpiContext<'a, 'b, 'c, 'info, SetAuthority<'info>> {
+        let cpi_accounts = SetAuthority {
+            account_or_mint: accounts
+                .initializer_deposit_token_account
+                .to_account_info()
+                .clone(),
+            current_authority: accounts.initializer.clone(),
+        };
+        let cpi_program = accounts.token_program.clone();
+        CpiContext::new(cpi_program, cpi_accounts)
+    }
+}
+
+impl<'a, 'b, 'c, 'info> CancelEscrow<'info> {
+    fn into_set_authority_context(
+        &self,
+        signer: &'a [&'b [&'c [u8]]],
```

Comment:
Tip. You can avoid declaring these lifetimes by removing the `signer` argument, using `CpiContext::new()`, and then having the caller use `.with_signer(&[seeds])`.